cypress-io/cypress
# feat: add logic for "cypress-triggered events"

- Closes #23765

### User facing changelog


No user facing changes.

### Additional details


This PR adds a utility function and store to support it, which between them gather together the logic needed to drive features related to the App and Cloud Integration. Much of the work was originally done on the branch for #23762, which will merge _after_ this. We've split this out into a PR that closes #23765 to make it easier to review. That other draft PR is a good place to observe the impact of using this function, as it implements the Smart Banner and Login flows.

 A good place to look at the improvement in DX is the watcher in SpecsListBanners.vue on that branch, that manages the open/closed banner state. Not included in this PR, but as a reference: Before:

  ```ts
 watch(
  cloudData,
  ([cloudViewer, cachedUser, currentProject]) => {
    // Cached user covers state where we're authenticated but data isn't loaded yet
    const isLoggedIn = !!cachedUser?.id || !!cloudViewer?.id
    // Need to be able to tell whether the lack of `firstOrganization` means they don't have an org or whether it just hasn't loaded yet
    // Not having this check can cause a brief flicker of the 'Create Org' banner while org data is loading
    const isOrganizationLoaded = !!cloudViewer?.firstOrganization
    const isMemberOfOrganization = (cloudViewer?.firstOrganization?.nodes?.length ?? 0) > 0
    const isProjectConnected = !!currentProject?.projectId && currentProject.cloudProject?.__typename === 'CloudProject'
    const hasNoRecordedRuns = currentProject?.cloudProject?.__typename === 'CloudProject' && (currentProject.cloudProject?.runs?.nodes?.length ?? 0) === 0
    const hasFourDaysOfCypressUse = (Date.now() - currentProject?.savedState?.firstOpened) > interval('4 days')
    
    showRecordBanner.value = !hasBannerBeenDismissed(BannerIds.ACI_082022_RECORD) && isLoggedIn && isProjectConnected && isMemberOfOrganization && isProjectConnected && hasNoRecordedRuns && hasFourDaysOfCypressUse
    showConnectBanner.value = !hasBannerBeenDismissed(BannerIds.ACI_082022_CONNECT_PROJECT) && isLoggedIn && isMemberOfOrganization && !isProjectConnected && hasFourDaysOfCypressUse
    showCreateOrganizationBanner.value = !hasBannerBeenDismissed(BannerIds.ACI_082022_CREATE_ORG) && isLoggedIn && isOrganizationLoaded && !isMemberOfOrganization && hasFourDaysOfCypressUse
    showLoginBanner.value = !hasBannerBeenDismissed(BannerIds.ACI_082022_LOGIN) && !isLoggedIn && hasFourDaysOfCypressUse
    
    hasRecordBannerBeenShown.value = hasBannerBeenShown(BannerIds.ACI_082022_RECORD)
    hasConnectBannerBeenShown.value = hasBannerBeenShown(BannerIds.ACI_082022_CONNECT_PROJECT)
    hasCreateOrganizationBannerBeenShown.value = hasBannerBeenShown(BannerIds.ACI_082022_CREATE_ORG)
    hasLoginBannerBeenShown.value = hasBannerBeenShown(BannerIds.ACI_082022_LOGIN)
  },
  { immediate: true },)
```

After: 

```ts
watch(cloudData, () => {
  if (!isAllowedFeature('specsListBanner')) {
    return
  }

  showRecordBanner.value = userStatusMatches('needsRecordedRun')
  showConnectBanner.value = userStatusMatches('needsProjectConnect')
  showCreateOrganizationBanner.value = userStatusMatches('needsOrgConnect')
  showLoginBanner.value = userStatusMatches('isLoggedOut')

  hasRecordBannerBeenShown.value = hasBannerBeenShown(BannerIds.ACI_082022_RECORD)
  hasConnectBannerBeenShown.value = hasBannerBeenShown(BannerIds.ACI_082022_CONNECT_PROJECT)
  hasCreateOrganizationBannerBeenShown.value = hasBannerBeenShown(BannerIds.ACI_082022_CREATE_ORG)
  hasLoginBannerBeenShown.value = hasBannerBeenShown(BannerIds.ACI_082022_LOGIN)
},
{ immediate: true })
```

The end result is we can now ask a helper function "Is the X allowed?" and have that helper provide the answer based on the user (and project's) progression through the "login -> connect -> record" journey (defined in the store's `userStatus` getter) and the other various constraints around timing that are defined in the rules for that feature as part of `isAllowedFeature.ts`.

On that implementation branch (#23762), the store is populated through a component that watches graphql for updates to the required field. The reason to forward those values into a client side store is to make testing a little more straightforward and allow components throughout both App and Launchpad to access these values as needed without adding fields to graphql. The store in this PR also includes some values like `isLoginConnectOpen` that support new modal behavior not specifically addressed in this PR that that are used in that branch.

There is also a little extra abstraction going on in the `rules` object of `isAllowedFeature`. The goal is to be able to clearly see in that section what are the rules under which Cypress will, say, show a banner, for a given user status. Between these rules and the userStatus getter in the store, we have two places where all of the lines and notes in this whimsical are managed: https://whimsical.com/get-to-first-run-TgixRWciqE1chPGDCHk5b2 - meaning understanding and maintaining these rules, and driving different features off of this core set of data, should require less work in future. 

### Steps to test


Open the unit test for the helper: `packages/frontend-shared/src/utils/isAllowedFeature.cy.ts`. Run the spec and see that all combinations are passing.

In `isAllowedFeature.ts`, change some rules. Setting something like `minTimeSinceEvent(events.cypressFirstOpened, '40 days')` on a feature should cause some related tests to fail based on the starting value we set of "5 days before now" for `firstOpened`. 

### Feedback sought


I'm especially curious if this is considered readable by others, and any other feedback is welcome too of course.

### PR Tasks


- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
